File: C++/task-scheduler.cpp
// Time:  O(n)
// Space: O(26) = O(1)

class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        unordered_map<char,int> count;
        int max_count = 0;
        for (const auto& task : tasks) {
            ++count[task];
            max_count = max(max_count, count[task]);
        }
        
        auto result = (max_count - 1) * (n + 1);
        for (const auto& kvp : count) {
            if (kvp.second == max_count) {
                ++result;
            }
        }
        return max(result, static_cast<int>(tasks.size()));
    }
};

File: Python/task-scheduler.py
# Time:  O(n)
# Space: O(26) = O(1)

from collections import Counter


class Solution(object):
    def leastInterval(self, tasks, n):
        """
        :type tasks: List[str]
        :type n: int
        :rtype: int
        """
        counter = Counter(tasks)
        _, max_count = counter.most_common(1)[0]
        return max((max_count-1) * (n+1) + counter.values().count(max_count), len(tasks))
